### PR TITLE
feat: use MC media-error-dialog

### PR DESCRIPTION
**Pull request by luwes**
related #1012
requires https://github.com/muxinc/media-chrome/pull/1024

- fixes an issue where the `.error` object could be null when the error event is dispatched.
this was introduced when we started calling initialize() function deferred in mux-video I believe.
this is fixed by saving the error state each time before we dispatch the error event.
- the warning API is not implemented because after closer inspection we don't really do much with non-fatal errors.
  the only thing is doing a console.warn here:
  https://github.com/muxinc/elements/blob/1e1f24a18d7fe61c5b4bdfd9cf4cfb6d9d57b0ea/packages/mux-player/src/index.ts#L461-L465
we can easily do that in playback-core if we want?

### CODE REVIEW

File: packages/mux-player/src/index.ts [hunk 1]
```diff
@@ -179,6 +179,34 @@ function getProps(el: MuxPlayerElement, state?: any): MuxTemplateProps {
   return props;
 }
 
+MediaErrorDialog.formatErrorMessage = (error: { code?: number; message?: string }) => {
```
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:182]: question: any tradeoffs you see on subclassing `MediaErrorDialog` with a `formatErrorMessage` vs. modifying `MediaErrorDialog`'s `formatErrorMessage`? One pro of this approach: we automatically get granular error messaging for all themes. One con of this approach: on the off chance that folks have `<mux-player>` and some other player on the same page, folks might get unexpected results. I suspect any perf etc. considerations are negligible (only a small amount of addititonal code required for subclassing). Any others on either side you can think of?

File: packages/mux-player/src/index.ts [hunk 2]
```diff
@@ -458,7 +493,7 @@ class MuxPlayerElement extends VideoApiElement implements MuxPlayerElement {
         return;
       }
 
-      const { dialog, devlog } = getErrorLogs(error, false);
+      const { devlog } = getErrorLogs(error, false);
```
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:496]: nitpick: the `dialog` and `devlog` code are already broken up into separate functions in the `errors.ts` module. Since their usage is no longer colocated, we should just export and use those directly (feel free to rename given their current verbosity).

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
Some questions/callouts on the draft

File: packages/mux-player/src/index.ts [hunk 3]
```diff
@@ -288,10 +315,18 @@ class MuxPlayerElement extends VideoApiElement implements MuxPlayerElement {
   #hotkeys = new AttributeTokenList(this, 'hotkeys');
   #state: Partial<MuxTemplateProps> = {
     ...initialState,
-    onCloseErrorDialog: () => this.#setState({ dialog: undefined, isDialogOpen: false }),
-    onInitFocusDialog: (e) => {
+    onCloseErrorDialog: (event) => {
```
[the remark this answers is not in the document]
Comment (luwes) [reply packages/mux-player/src/index.ts:318]: it is possible but we use the `isDialogOpen` state to change 2 attributes in the theme; `disabled` and `nohotkeys`.

enabling `nohotkeys` seems an easy one to handle in Media Chrome only if the dialog is open.

setting `disabled` on all controls directly when the dialog opens feels forceful.
thoughts?

[the remark this answers is not in the document]
Comment (luwes) [reply packages/mux-player/src/index.ts:318]: resolve: we disable all state change requests when a fatal error occurs.

adding `disabled` or `inert` to UI controls to disable keyboard focus is for now left up to the consumer.

File: packages/mux-player/src/index.ts (hunk 1 above)
Comment (luwes) [reply packages/mux-player/src/index.ts:182]: we put a pin in this. we will go forward with this static override for now but maybe this could be avoided by having the end user message come from the custom media element directly?

Comment (cjpillsbury) [reply packages/mux-player/src/index.ts:182]: ah we (or at least I) actually said we would be mostly fine with this approach with one small addendum - in the `else` case, we would use the base impl of `formatErrorMessage()`. Something like:

```ts
const baseFormatErrorMessage = MediaErrorDialog.formatErrorMessage;
MediaErrorDialog.formatErrorMessage = (error: { code?: number; message?: string }) => {
  if (error instanceof MediaError) {
    // ...
  }
  return baseFormatErrorMessage(error);
};
```

File: packages/mux-player/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjpillsbury) [reply packages/mux-player/src/index.ts:318]: Fine for now but I think we should be able to remove most/all of this as a followup either here and/or in media chrome

File: packages/mux-player/src/themes/classic/classic.html [hunk 4]
```diff
@@ -521,6 +521,7 @@
     <slot name="media" slot="media"></slot>
     <slot name="poster" slot="poster"></slot>
     <media-loading-indicator slot="centered-chrome" noautohide></media-loading-indicator>
+    <media-error-dialog noautohide></media-error-dialog>
```
Comment (cjpillsbury) [inline packages/mux-player/src/themes/classic/classic.html:524]: nitpick(blocking): `slot="dialog"`

File: packages/mux-player/tsconfig.json [hunk 5]
```diff
@@ -18,7 +18,7 @@
     "esModuleInterop": true,
     "skipLibCheck": false,
     "forceConsistentCasingInFileNames": true,
-    "typeRoots": ["node_modules/@types", "../../types/media-chrome.d.ts"]
```
Comment (cjpillsbury) [inline packages/mux-player/tsconfig.json:21]: question(blocking): Why were these types things changed for this PR? This might be right (I'd want to make sure for all the different use cases we have), but unless this is needed for the PR, these changes should be backed out and added to a separate PR.

File: packages/playback-core/package.json [hunk 6]
```diff
@@ -51,7 +51,7 @@
     "build:esm-module": "esbuilder src/index.ts --sourcemap --format=esm --outfile=./dist/playback-core.mjs",
     "build:cjs": "esbuilder src/index.ts --sourcemap --format=cjs --out-extension:.js=.cjs.js --external:mux-embed --external:hls.js",
     "build:iife": "esbuilder src/index.ts --sourcemap --format=iife --outfile=./dist/playback-core.js",
-    "prebuild:types": "shx mkdir -p ./dist/types && shx cp ../../types/** ./dist/types/",
```
Comment (cjpillsbury) [inline packages/playback-core/package.json:54]: Same as question, above.

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
some small callouts and looks like there are some linting errors, but this is def close.

File: packages/mux-player/src/index.ts (hunk 1 above)
Comment (luwes) [reply packages/mux-player/src/index.ts:182]: yes, good catch. I somehow thought it already did that.

File: packages/mux-player/tsconfig.json (hunk 5 above)
Comment (luwes) [reply packages/mux-player/tsconfig.json:21]: this was needed to make it build, because the MediaErrorDialog class is imported.

the media-chrome.d.ts file was excluding the types that are now available in the Media Chrome pkg itself

```ts
declare module 'media-chrome' {
  export class MediaTheme extends HTMLElement {}
  export class MediaController extends HTMLElement {
    media: HTMLVideoElement;
  }
}
```

**Comment by luwes** [comment]
test page loading a 404 source 
https://elements-demo-vanilla-git-fork-luwes-media-error-dialog-mux.vercel.app/mux-player-error.html
